usr/lib/pkcs11/ep11_stdll/ep11_specific.c

```diff
@@ -7692,3 +7706,194 @@ CK_BBOOL ep11tok_optimize_single_ops(STDLL_TokData_t *tokdata)
+    if (rc != CKR_OK)
+        return rc;
+    if (sscanf(buf, "CEX%luP", type) != 1)
+        return CKR_FUNCTION_FAILED;
+    return CKR_OK;
+}
+
+typedef struct query_version
+{
+    CK_ULONG firmware_API_version;
+    CK_VERSION firmware_version;
+    CK_ULONG card_type;
+    CK_BBOOL first;
+    CK_BBOOL error;
+} query_version_t;
+
+static CK_RV version_query_handler(uint_32 adapter, uint_32 domain,
+                                   void *handler_data)
+{
+    query_version_t *qv = (query_version_t *)handler_data;
+    CK_IBM_XCP_INFO xcp_info;
+    CK_ULONG xcp_info_len = sizeof(xcp_info);
+    CK_RV rc;
+    ep11_target_t target;
+    CK_ULONG card_type;
+
+    memset(&target, 0, sizeof(target));
+    target.length = 1;
+    target.apqns[0] = adapter;
+    target.apqns[1] = domain;
+
+    rc = dll_m_get_xcp_info(&xcp_info, &xcp_info_len, CK_IBM_XCPQ_MODULE, 0,
+                           (uint64_t)&target);
+   if (rc != CKR_OK) {
+       TRACE_ERROR("%s Failed to query module version from adapter %02X.%04X\n",
+                           __func__, adapter, domain);
+       /* card may no longer be online, so ignore this error situation */
+       return CKR_OK;
+   }
+
+   rc = get_card_type(adapter, &card_type);
+   if (rc != CKR_OK) {
+       TRACE_ERROR("%s Failed to get card type for adapter %02X.%04X\n",
+                   __func__, adapter, domain);
+       /* card may no longer be online, so ignore this error situation */
+       return CKR_OK;
+   }
+
+   if (!qv->first && qv->firmware_API_version != xcp_info.firmwareApi) {
+       TRACE_ERROR("%s Adapter %02X.%04X has a different API version "
+                                 "than the previous adapters: %lu\n",
+                                 __func__, adapter, domain,
+                                 xcp_info.firmwareApi);
+       OCK_SYSLOG(LOG_ERR,
+                  "Warning: Adapter %02X.%04X has a different API version "
+                  " than the previous adapters: %lu\n",
+                  adapter, domain, xcp_info.firmwareApi);
+       qv->error = TRUE;
+       return CKR_OK;
+   }
+
+   if (!qv->first && compare_ck_version(&qv->firmware_version,
+                                        &xcp_info.firmwareVersion) != 0) {
+       TRACE_ERROR("%s Adapter %02X.%04X has a different firmware version"
+                                 "than the previous adapters: %d.%d\n",
+                                 __func__, adapter, domain,
+                                 xcp_info.firmwareVersion.major,
+                                 xcp_info.firmwareVersion.minor);
+       OCK_SYSLOG(LOG_ERR,
+                  "Warning: Adapter %02X.%04X has a different firmware version"
+                  " than the previous adapters: %d.%d\n",
+                  adapter, domain,  xcp_info.firmwareVersion.major,
+                  xcp_info.firmwareVersion.minor);
+       qv->error = TRUE;
+       return CKR_OK;
+   }
+
+   if (!qv->first && qv->card_type != card_type) {
```

Comment:
Would't it make sense to move the card type check before the firmware version check? I assume that for different card types the firmware is anyway different.

Reply:
Yes good point. Will move the card-type check as first check.